### PR TITLE
*: use a key to decide mdl is enabled or not

### PR DESCRIPTION
**Pull request by wjhuang2016**
Signed-off-by: wjhuang2016 <huangwenjun1997@gmail.com>

### What problem does this PR solve?

Use a key to decide mdl is enabled or not.
### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
upgrade from v6.1 v6.2 v6.3
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: session/bootstrap.go [hunk 1]
```diff
@@ -867,12 +872,19 @@ func upgrade(s Session) {
...
+	if isNull {
+		upgradeToVer99Before(s)
+	}
 	for _, upgrade := range bootstrapVersion {
 		upgrade(s, ver)
 	}
-	if needEnableMdl {
-		upgradeToVer99After(s, ver)
+	if isNull {
+		upgradeToVer99After(s)
```
Comment (lcwangchao) [inline session/bootstrap.go:887]: Where is setting metalock flag in meta after upgrade?

**Review by lcwangchao — APPROVED** [review]

File: ddl/ddl.go [hunk 2]
```diff
@@ -1259,6 +1259,20 @@ func (d *ddl) SwitchMDL(enable bool) error {
...
 
 	variable.EnableMDL.Store(enable)
+	err = kv.RunInNewTxn(kv.WithInternalSourceType(context.Background(), kv.InternalTxnDDL), d.store, true, func(ctx context.Context, txn kv.Transaction) error {
+		m := meta.NewMeta(txn)
+		oldEnable, _, err := m.GetMetadataLock()
+		if err != nil {
+			return err
+		}
+		if oldEnable != enable {
+			err = meta.NewMeta(txn).SetMetadataLock(enable)
```
Comment (tangenta) [inline ddl/ddl.go:1269]:
```suggestion
			err = m.SetMetadataLock(enable)
```

File: session/bootstrap.go [hunk 3]
```diff
@@ -867,12 +873,19 @@ func upgrade(s Session) {
 		}
 	}
 	// Do upgrade works then update bootstrap version.
-	needEnableMdl := upgradeToVer99Before(s, ver)
+	isNull, err := InitMDLVariableForUpgrade(s.GetStore())
+	if err != nil {
+		logutil.BgLogger().Fatal("[Upgrade] init metadata lock failed", zap.Error(err))
```
Comment (tangenta) [inline session/bootstrap.go:878]:
```suggestion
		logutil.BgLogger().Fatal("[upgrade] init metadata lock failed", zap.Error(err))
```

File: session/session.go [hunk 4]
```diff
@@ -2848,6 +2848,64 @@ func InitMDLTable(store kv.Storage) error {
...
+		return nil
+	})
+	if isNull {
+		return errors.New("metadata lock is null")
+	}
+	if !enable {
+		variable.EnableMDL.Store(false)
+	} else {
+		variable.EnableMDL.Store(true)
+	}
```
Comment (tangenta) [inline session/session.go:2905]:
```suggestion
	variable.EnableMDL.Store(enable)
```

**Review by XuHuaiyu — APPROVED** [review]

**Pull request merged**